src/java.base/share/classes/jdk/internal/classfile/impl/verifier/ParserVerifier.java

```diff
@@ -0,0 +1,371 @@
+                }
+                case MethodRefEntry mre -> {
+                    check.accept(mre.owner()::asSymbol);
+                    check.accept(mre::typeSymbol);
+                    yield () -> verifyMethodName(mre.name().stringValue());
+                }
+                case ModuleEntry me -> me::asSymbol;
+                case NameAndTypeEntry nate -> {
+                    check.accept(nate.name()::stringValue);
+                    yield () -> nate.type().stringValue();
+                }
+                case PackageEntry pe -> pe::asSymbol;
+            });
+        }
+    }
+
+    private void verifyFieldName(String name) {
+        if (name.length() == 0 || name.chars().anyMatch(ch -> switch(ch) {
+                    case '.', ';', '[', '/' -> true;
+                    default -> false;
+                })) {
+              throw new VerifyError("Illegal field name %s in %s".formatted(name, toString(classModel)));
+        }
+    }
+
+    private void verifyMethodName(String name) {
+        if (!name.equals(INIT_NAME)
+            && !name.equals(CLASS_INIT_NAME)
+            && (name.length() == 0 || name.chars().anyMatch(ch -> switch(ch) {
+                    case '.', ';', '[', '/', '<', '>' -> true;
+                    default -> false;
+                }))) {
+              throw new VerifyError("Illegal method name %s in %s".formatted(name, toString(classModel)));
+        }
+    }
+
+    private void verifyInterfaces(List<VerifyError> errors) {
+        var intfs = new HashSet<ClassEntry>();
+        for (var intf : classModel.interfaces()) {
+            if (!intfs.add(intf)) {
+                errors.add(new VerifyError("Duplicate interface %s in %s".formatted(intf.asSymbol().displayName(), toString(classModel))));
+            }
+        }
+    }
+
+    private void verifyFields(List<VerifyError> errors) {
+        var fields = new HashSet<String>();
+        for (var f : classModel.fields()) try {
+            if (!fields.add(f.fieldName().stringValue() + f.fieldType().stringValue())) {
+                errors.add(new VerifyError("Duplicate field name %s with signature %s in %s".formatted(f.fieldName().stringValue(), f.fieldType().stringValue(), toString(classModel))));
+            }
+            verifyFieldName(f.fieldName().stringValue());
+        } catch (VerifyError ve) {
+            errors.add(ve);
+        }
+    }
+
+    private void verifyMethods(List<VerifyError> errors) {
+        var methods = new HashSet<String>();
+        for (var m : classModel.methods()) try {
+            if (!methods.add(m.methodName().stringValue() + m.methodType().stringValue())) {
+                errors.add(new VerifyError("Duplicate method name %s with signature %s in %s".formatted(m.methodName().stringValue(), m.methodType().stringValue(), toString(classModel))));
+            }
+            if (m.methodName().equalsString(CLASS_INIT_NAME)
+                    && !m.flags().has(AccessFlag.STATIC)) {
```

Comment:
Do we verify it has void return and (since class file version JAVA 7) takes no args? The static requirement is since JAVA 7 too.

Reply:
This is a first round of transcription from `classFileParser.cpp` and I most probably missed many verifications applied there. Feel free to point to a specific verification code in the `classFileParser.cpp` I missed, however to avoid very complex synchronization I preferred to do not invent a custom list of verifications. Relevant parts are widely spread across the specifications and collecting and tracking them would be very complex task.